File: clients/go/internal/delve/manager.go
// Package delve provides Delve process lifecycle management.
package delve

import (
	"bytes"
	"errors"
	"fmt"
	"log/slog"
	"net"
	"os"
	"os/exec"
	"strconv"
	"strings"
	"syscall"
	"time"
)

// maxPortRetries is the number of times to retry port allocation
// when the ephemeral port is taken between allocation and Delve startup
const maxPortRetries = 3

// Process represents a running Delve process.
type Process struct {
	Cmd    *exec.Cmd
	Host   string
	Port   int
	stdout *bytes.Buffer
	stderr *bytes.Buffer
}

// Logs returns the combined stdout and stderr output from Delve.
func (p *Process) Logs() string {
	if p == nil {
		return ""
	}
	var result string
	if p.stdout != nil {
		result += "=== STDOUT ===\n" + p.stdout.String()
	}
	if p.stderr != nil {
		result += "\n=== STDERR ===\n" + p.stderr.String()
	}
	return result
}

// Manager handles Delve process lifecycle.
type Manager struct {
	DelvePath string
	Timeout   time.Duration
}

// NewManager creates a new Delve manager.
func NewManager(delvePath string, timeout time.Duration) *Manager {
	if delvePath == "" {
		delvePath = "dlv" // Use PATH lookup
	}
	if timeout == 0 {
		timeout = 10 * time.Second
	}
	return &Manager{
		DelvePath: delvePath,
		Timeout:   timeout,
	}
}

// FindDelve locates the Delve binary.
// Returns the path if found, or an error if not found.
func FindDelve(delvePath string) (string, error) {
	if delvePath != "" {
		// Check if explicit path exists
		if _, err := os.Stat(delvePath); err == nil {
			return delvePath, nil
		}
		return "", fmt.Errorf("delve not found at %s", delvePath)
	}

	// Search in PATH
	path, err := exec.LookPath("dlv")
	if err != nil {
		return "", fmt.Errorf("delve (dlv) not found in PATH: %w", err)
	}
	return path, nil
}

// SpawnAndAttach spawns Delve to attach to the current process.
// It waits for the DAP server to become ready before returning.
//
// When port is 0, an ephemeral port is allocated. Due to a TOCTOU race
// (the port may be taken between allocation and Delve startup), this
// operation is retried up to maxPortRetries times.
func (m *Manager) SpawnAndAttach(host string, port int) (*Process, error) {
	// Find delve binary
	delvePath, err := FindDelve(m.DelvePath)
	if err != nil {
		return nil, err
	}

	// If port is specified (non-zero), no retries needed
	if port != 0 {
		return m.spawnDelve(delvePath, host, port)
	}

	// Port 0: ephemeral port allocation with retry logic
	// This handles the TOCTOU race where the port may be taken between
	// net.Listen close and Delve startup
	var lastErr error
	for attempt := 0; attempt < maxPortRetries; attempt++ {
		// Allocate an ephemeral port
		listener, err := net.Listen("tcp", fmt.Sprintf("%s:0", host))
		if err != nil {
			return nil, fmt.Errorf("failed to find available port: %w", err)
		}
		actualPort := listener.Addr().(*net.TCPAddr).Port
		if err := listener.Close(); err != nil {
			slog.Debug("failed to close port listener", "error", err)
		} // Release port for Delve to bind

		proc, err := m.spawnDelve(delvePath, host, actualPort)
		if err == nil {
			return proc, nil
		}

		// Check if it's a port bind error (TOCTOU race)
		if isPortBindError(err) {
			lastErr = err
			continue // Retry with a new port
		}

		// Non-port error, fail immediately
		return nil, err
	}

	return nil, fmt.Errorf("failed after %d port allocation retries: %w", maxPortRetries, lastErr)
}

// spawnDelve spawns a Delve process on the specified port.
func (m *Manager) spawnDelve(delvePath string, host string, port int) (*Process, error) {
	pid := os.Getpid()

	// Build command:
	// dlv attach <pid> --headless --accept-multiclient --api-version=2 --continue --listen=<host>:<port>
	// The --continue flag is critical: without it, Delve pauses the process on attach,
	// which would hang the wake request (the process can't respond while paused).
	args := []string{
		"attach",
		strconv.Itoa(pid),
		"--headless",
		"--accept-multiclient",
		"--api-version=2",
		"--continue",
		fmt.Sprintf("--listen=%s:%d", host, port),
	}

	cmd := exec.Command(delvePath, args...)

	// Capture output for debugging
	var stdout, stderr bytes.Buffer
	cmd.Stdout = &stdout
	cmd.Stderr = &stderr

	// Set process group so we can kill the whole group later
	cmd.SysProcAttr = &syscall.SysProcAttr{
		Setpgid: true,
	}

	if err := cmd.Start(); err != nil {
		return nil, fmt.Errorf("failed to start delve: %w", err)
	}

	proc := &Process{
		Cmd:    cmd,
		Host:   host,
		Port:   port,
		stdout: &stdout,
		stderr: &stderr,
	}

	// Wait for DAP port to accept connections
	if err := m.waitForReady(host, port, cmd.Process); err != nil {
		// Kill process if startup failed
		if killErr := m.Kill(proc); killErr != nil {
			slog.Warn("failed to kill delve process during cleanup", "error", killErr)
		}
		return nil, fmt.Errorf("delve failed to start: %w\nLogs:\n%s", err, proc.Logs())
	}

	return proc, nil
}

// isPortBindError checks if the error indicates a port bind failure
// (address already in use), which suggests a TOCTOU race condition.
func isPortBindError(err error) bool {
	if err == nil {
		return false
	}
	errStr := err.Error()
	// Common indicators of port bind failure
	return strings.Contains(errStr, "address already in use") ||
		strings.Contains(errStr, "bind: address already in use") ||
		errors.Is(err, syscall.EADDRINUSE)
}

// waitForReady polls the DAP port until it accepts connections.
func (m *Manager) waitForReady(host string, port int, proc *os.Process) error {
	deadline := time.Now().Add(m.Timeout)
	addr := net.JoinHostPort(host, strconv.Itoa(port))
	checkInterval := 100 * time.Millisecond

	for time.Now().Before(deadline) {
		// Check if process died
		if !processAlive(proc) {
			return fmt.Errorf("delve process terminated unexpectedly")
		}

		// Try to connect
		conn, err := net.DialTimeout("tcp", addr, checkInterval)
		if err == nil {
			if err := conn.Close(); err != nil {
				slog.Debug("failed to close probe connection", "error", err)
			}
			return nil
		}

		time.Sleep(checkInterval)
	}

	return fmt.Errorf("delve did not become ready within %v", m.Timeout)
}

// processAlive checks if a process is still running.
func processAlive(proc *os.Process) bool {
	if proc == nil {
		return false
	}
	// Signal 0 checks if process exists without sending a signal
	err := proc.Signal(syscall.Signal(0))
	return err == nil
}

// Kill terminates the Delve process gracefully.
func (m *Manager) Kill(proc *Process) error {
	if proc == nil || proc.Cmd == nil || proc.Cmd.Process == nil {
		return nil
	}

	// Try graceful shutdown first (SIGTERM)
	if err := proc.Cmd.Process.Signal(syscall.SIGTERM); err != nil {
		// Process might already be dead
		return nil
	}

	// Wait with timeout
	done := make(chan error, 1)
	go func() {
		_, err := proc.Cmd.Process.Wait()
		done <- err
	}()

	select {
	case <-done:
		return nil
	case <-time.After(2 * time.Second):
		// Force kill
		if err := proc.Cmd.Process.Kill(); err != nil {
			slog.Warn("delve force kill failed", "error", err)
		}
		// Wait for it to actually die
		<-done
		return nil
	}
}

File: clients/rust/src/lldb/manager.rs
//! LLDB process lifecycle management.

use std::io::{BufRead, BufReader};
use std::net::TcpStream;
use std::path::PathBuf;
use std::process::{Child, Command, Stdio};
use std::time::{Duration, Instant};

use tracing::{debug, trace, warn};

use crate::error::{Error, Result};

/// Maximum number of port allocation retries.
const MAX_PORT_RETRIES: u32 = 3;

/// Information about a running lldb-dap process.
#[derive(Debug)]
pub struct LldbProcess {
    /// Child process handle.
    pub child: Child,

    /// Host the DAP server is listening on.
    #[allow(dead_code)]
    pub host: String,

    /// Port the DAP server is listening on.
    pub port: u16,
}

/// Manager for lldb-dap process lifecycle.
pub struct LldbManager {
    /// Path to lldb-dap binary.
    lldb_dap_path: PathBuf,

    /// Timeout for lldb-dap to start.
    timeout: Duration,
}

impl LldbManager {
    /// Create a new LLDB manager.
    pub fn new(lldb_dap_path: PathBuf, timeout: Duration) -> Self {
        Self {
            lldb_dap_path,
            timeout,
        }
    }

    /// Spawn lldb-dap and attach to the current process.
    ///
    /// When port is 0, an ephemeral port is allocated. Due to a TOCTOU race
    /// (the port may be taken between allocation and lldb-dap startup), this
    /// operation is retried up to MAX_PORT_RETRIES times.
    pub fn spawn_and_attach(&self, host: &str, port: u16) -> Result<LldbProcess> {
        // If port is specified (non-zero), no retries needed
        if port != 0 {
            return self.spawn_lldb(host, port);
        }

        // Port 0: ephemeral port allocation with retry logic
        let mut last_err = None;
        for attempt in 0..MAX_PORT_RETRIES {
            // Allocate an ephemeral port
            let actual_port = allocate_port(host)?;
            debug!(
                "Attempt {}: allocated ephemeral port {}",
                attempt + 1,
                actual_port
            );

            match self.spawn_lldb(host, actual_port) {
                Ok(process) => return Ok(process),
                Err(e) => {
                    if is_port_bind_error(&e) {
                        warn!("Port {} taken (TOCTOU race), retrying...", actual_port);
                        last_err = Some(e);
                        continue;
                    }
                    return Err(e);
                }
            }
        }

        Err(last_err
            .unwrap_or_else(|| Error::LldbStartFailed("failed after max port retries".to_string())))
    }

    /// Spawn lldb-dap on the specified port.
    ///
    /// IMPORTANT: We do NOT send the DAP attach request here! The Detrix daemon
    /// (a separate process) will connect to lldb-dap and send the attach request.
    /// If we tried to send attach from within this process, lldb-dap would use
    /// ptrace to stop us, causing a deadlock.
    fn spawn_lldb(&self, host: &str, port: u16) -> Result<LldbProcess> {
        // Build command:
        // lldb-dap --connection listen://host:port
        let listen_addr = format!("listen://{}:{}", host, port);

        debug!(
            "Starting lldb-dap: {:?} --connection {}",
            self.lldb_dap_path, listen_addr
        );

        let mut child = Command::new(&self.lldb_dap_path)
            .args(["--connection", &listen_addr])
            .stdin(Stdio::null())
            .stdout(Stdio::piped())
            .stderr(Stdio::piped())
            .spawn()
            .map_err(|e| Error::LldbStartFailed(e.to_string()))?;

        // Wait for lldb-dap to accept connections
        if let Err(e) = self.wait_for_ready(host, port, &mut child) {
            // Kill process if startup failed
            let _ = self.kill_process(&mut child);
            return Err(e);
        }

        debug!(
            "lldb-dap ready on {}:{}, daemon will attach to PID {}",
            host,
            port,
            std::process::id()
        );

        // NOTE: We do NOT send DAP initialize/attach here.
        // The Detrix daemon will connect and handle the DAP protocol.
        // This avoids a deadlock where lldb-dap would stop this process
        // via ptrace while we're waiting for its response.

        Ok(LldbProcess {
            child,
            host: host.to_string(),
            port,
        })
    }

    /// Wait for lldb-dap to accept connections.
    fn wait_for_ready(&self, host: &str, port: u16, child: &mut Child) -> Result<()> {
        let deadline = Instant::now() + self.timeout;
        let check_interval = Duration::from_millis(100);

        while Instant::now() < deadline {
            // Check if process died
            if let Ok(Some(status)) = child.try_wait() {
                // Read stderr for error message
                let stderr = child
                    .stderr
                    .take()
                    .map(|s| {
                        BufReader::new(s)
                            .lines()
                            .take(10)
                            .filter_map(|l| l.ok())
                            .collect::<Vec<_>>()
                            .join("\n")
                    })
                    .unwrap_or_default();

                return Err(Error::LldbStartFailed(format!(
                    "lldb-dap exited with status {}: {}",
                    status, stderr
                )));
            }

            // Try to connect
            let addr = format!("{}:{}", host, port);
            if TcpStream::connect_timeout(
                &addr
                    .parse()
                    .map_err(|e| Error::LldbStartFailed(format!("invalid address: {}", e)))?,
                check_interval,
            )
            .is_ok()
            {
                trace!("lldb-dap accepting connections on {}", addr);
                return Ok(());
            }

            std::thread::sleep(check_interval);
        }

        Err(Error::Timeout(format!(
            "lldb-dap did not become ready within {:?}",
            self.timeout
        )))
    }

    /// Send DAP initialize and attach requests.
    ///
    /// NOTE: This is kept for potential standalone testing, but is NOT used
    /// in normal operation. The Detrix daemon handles the DAP protocol.
    #[allow(dead_code)]
    fn send_attach_request(&self, host: &str, port: u16, pid: u32) -> Result<()> {
        let addr = format!("{}:{}", host, port);
        debug!("Connecting to lldb-dap at {}", addr);
        let mut stream = TcpStream::connect(&addr)
            .map_err(|e| Error::LldbStartFailed(format!("failed to connect to lldb-dap: {}", e)))?;
        debug!("Connected to lldb-dap");

        stream.set_read_timeout(Some(Duration::from_secs(5))).ok();
        stream.set_write_timeout(Some(Duration::from_secs(5))).ok();

        // Send DAP initialize request
        let init_request = serde_json::json!({
            "seq": 1,
            "type": "request",
            "command": "initialize",
            "arguments": {
                "clientID": "detrix-rust-client",
                "clientName": "Detrix Rust Client",
                "adapterID": "lldb-dap",
                "pathFormat": "path",
                "linesStartAt1": true,
                "columnsStartAt1": true,
                "supportsVariableType": true,
                "supportsVariablePaging": true,
                "supportsRunInTerminalRequest": false,
                "locale": "en-US"
            }
        });

        debug!("Sending initialize request");
        send_dap_message(&mut stream, &init_request)?;

        // Read initialize response
        debug!("Waiting for initialize response");
        let response = read_dap_response(&mut stream, "initialize")?;
        debug!("Initialize response: {:?}", response);

        if response.get("success") != Some(&serde_json::Value::Bool(true)) {
            let message = response
                .get("message")
                .and_then(|m| m.as_str())
                .unwrap_or("unknown error");
            return Err(Error::LldbStartFailed(format!(
                "initialize failed: {}",
                message
            )));
        }

        // Send attach request
        let attach_request = serde_json::json!({
            "seq": 2,
            "type": "request",
            "command": "attach",
            "arguments": {
                "pid": pid,
                "stopOnEntry": false
            }
        });

        debug!("Sending attach request for PID {}", pid);
        send_dap_message(&mut stream, &attach_request)?;

        // Read attach response - NOTE: lldb-dap may send events before the response
        debug!("Waiting for attach response (may receive events first)");
        let response = read_dap_response(&mut stream, "attach")?;
        debug!("Attach response: {:?}", response);

        if response.get("success") != Some(&serde_json::Value::Bool(true)) {
            let message = response
                .get("message")
                .and_then(|m| m.as_str())
                .unwrap_or("unknown error");
            return Err(Error::LldbStartFailed(format!(
                "attach failed: {}",
                message
            )));
        }

        // Send configurationDone request
        let config_done_request = serde_json::json!({
            "seq": 3,
            "type": "request",
            "command": "configurationDone",
            "arguments": {}
        });

        debug!("Sending configurationDone request");
        send_dap_message(&mut stream, &config_done_request)?;

        // Read configurationDone response
        debug!("Waiting for configurationDone response");
        let response = read_dap_response(&mut stream, "configurationDone")?;
        debug!("ConfigurationDone response: {:?}", response);

        // Keep the connection open for the daemon to use
        // The stream will be closed when it goes out of scope

        debug!("lldb-dap attached to PID {}", pid);
        Ok(())
    }

    /// Kill the lldb-dap process gracefully.
    pub fn kill(&self, process: &mut LldbProcess) -> Result<()> {
        self.kill_process(&mut process.child)
    }

    /// Kill a child process gracefully.
    fn kill_process(&self, child: &mut Child) -> Result<()> {
        #[cfg(unix)]
        {
            use nix::sys::signal::{kill, Signal};
            use nix::unistd::Pid;

            let pid = Pid::from_raw(child.id() as i32);

            // Try graceful shutdown first (SIGTERM)
            if kill(pid, Signal::SIGTERM).is_ok() {
                // Wait with timeout
                let deadline = Instant::now() + Duration::from_secs(2);
                while Instant::now() < deadline {
                    if child.try_wait().ok().flatten().is_some() {
                        return Ok(());
                    }
                    std::thread::sleep(Duration::from_millis(100));
                }
            }

            // Force kill
            let _ = child.kill();
            let _ = child.wait();
        }

        #[cfg(not(unix))]
        {
            let _ = child.kill();
            let _ = child.wait();
        }

        Ok(())
    }
}

/// Allocate an ephemeral port.
fn allocate_port(host: &str) -> Result<u16> {
    use std::net::TcpListener;

    let addr = format!("{}:0", host);
    let listener = TcpListener::bind(&addr)
        .map_err(|e| Error::PortBindError(format!("failed to bind to {}: {}", addr, e)))?;

    let port = listener
        .local_addr()
        .map_err(|e| Error::PortBindError(e.to_string()))?
        .port();

    // Drop listener to release port for lldb-dap
    drop(listener);

    Ok(port)
}

/// Check if the error indicates a port bind failure.
fn is_port_bind_error(err: &Error) -> bool {
    match err {
        Error::PortBindError(_) => true,
        Error::LldbStartFailed(msg) => {
            msg.contains("address already in use") || msg.contains("bind")
        }
        _ => false,
    }
}

/// Send a DAP message over the stream.
fn send_dap_message(stream: &mut TcpStream, message: &serde_json::Value) -> Result<()> {
    use std::io::Write;

    let body = serde_json::to_string(message)?;
    let header = format!("Content-Length: {}\r\n\r\n", body.len());

    stream
        .write_all(header.as_bytes())
        .map_err(|e| Error::LldbStartFailed(format!("failed to write header: {}", e)))?;
    stream
        .write_all(body.as_bytes())
        .map_err(|e| Error::LldbStartFailed(format!("failed to write body: {}", e)))?;
    stream
        .flush()
        .map_err(|e| Error::LldbStartFailed(format!("failed to flush: {}", e)))?;

    Ok(())
}

/// Read a DAP response, skipping any events that come before it.
fn read_dap_response(stream: &mut TcpStream, expected_command: &str) -> Result<serde_json::Value> {
    let start = Instant::now();
    let timeout = Duration::from_secs(10);

    loop {
        if start.elapsed() > timeout {
            return Err(Error::Timeout(format!(
                "timed out waiting for {} response",
                expected_command
            )));
        }

        let msg = read_dap_message(stream)?;
        let msg_type = msg.get("type").and_then(|t| t.as_str()).unwrap_or("");

        match msg_type {
            "response" => {
                // This is what we're looking for
                return Ok(msg);
            }
            "event" => {
                // Skip events, but log them
                let event_name = msg.get("event").and_then(|e| e.as_str()).unwrap_or("?");
                eprintln!("[LLDB] Skipping event: {}", event_name);
                continue;
            }
            other => {
                eprintln!("[LLDB] Unexpected message type: {}", other);
                continue;
            }
        }
    }
}

/// Read a DAP message from the stream.
fn read_dap_message(stream: &mut TcpStream) -> Result<serde_json::Value> {
    use std::io::{BufRead, BufReader, Read};

    let mut reader = BufReader::new(
        stream
            .try_clone()
            .map_err(|e| Error::LldbStartFailed(format!("failed to clone stream: {}", e)))?,
    );

    // Read headers
    let mut content_length: Option<usize> = None;
    loop {
        let mut line = String::new();
        reader
            .read_line(&mut line)
            .map_err(|e| Error::LldbStartFailed(format!("failed to read header: {}", e)))?;

        let line = line.trim();
        if line.is_empty() {
            break;
        }

        if let Some(value) = line.strip_prefix("Content-Length:") {
            content_length = value.trim().parse().ok();
        }
    }

    let content_length = content_length
        .ok_or_else(|| Error::LldbStartFailed("missing Content-Length header".to_string()))?;

    // Read body
    let mut body = vec![0u8; content_length];
    reader
        .read_exact(&mut body)
        .map_err(|e| Error::LldbStartFailed(format!("failed to read body: {}", e)))?;

    serde_json::from_slice(&body)
        .map_err(|e| Error::LldbStartFailed(format!("failed to parse response: {}", e)))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_allocate_port() {
        let port = allocate_port("127.0.0.1").unwrap();
        assert!(port > 0);
    }

    #[test]
    fn test_is_port_bind_error() {
        assert!(is_port_bind_error(&Error::PortBindError(
            "test".to_string()
        )));
        assert!(is_port_bind_error(&Error::LldbStartFailed(
            "address already in use".to_string()
        )));
        assert!(!is_port_bind_error(&Error::LldbNotFound(
            "test".to_string()
        )));
    }
}
